File: 1405-longest-happy-string/1405-longest-happy-string.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        string res =""; //contain resultent string
        priority_queue<pair<int,char>>maxheap;  //to strore max freq on top of queue
        for(auto [cnt,ch] : {make_pair(a,'a'),make_pair(b,'b'),make_pair(c,'c')})  //stores as {count,char} = frequency of three char 
          if(cnt>0)maxheap.push({cnt,ch});  //eleminate those who have 0 freq

         while(!maxheap.empty()){   //travese till maxheap has char left inside
            auto [cnt_curr,ch_curr] = maxheap.top(); //stors currcount and currchar with max freq
            maxheap.pop();  // Remove it from the queue
    
            // If appending this character doesn't violate the rule of having > 2 consecutive characters
            if (res.size() < 2 || res.back() != ch_curr || res[res.size() - 2] != ch_curr) {
                res += ch_curr; // Append the character to the result
                cnt_curr--;     // Decrease its frequency
                if (cnt_curr > 0) {
                    maxheap.push({cnt_curr, ch_curr}); // Reinsert the character with updated frequency
                }
            } else {
                // If we can't append this character, we need to check the next most frequent character
                if (maxheap.empty()) break;  // No other characters available, we cannot proceed
                
                auto [cnt_next, ch_next] = maxheap.top();  // Get the next most frequent character
                maxheap.pop();  // Remove it from the queue
                
                res += ch_next; // Append this second character
                // Decrease its frequency
                if (--cnt_next > 0) {
                    maxheap.push({cnt_next, ch_next}); // Reinsert the second character with updated frequency
                }
                maxheap.push({cnt_curr, ch_curr});  // Reinsert the original character
            }
        }
        return res;
    }
};      

File: 1405-longest-happy-string/1405-longest-happy-string.py
import heapq

class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        res = []  # Resultant string as list of characters
        max_heap = []
        
        # Push available characters with their frequencies
        if a > 0:
            heapq.heappush(max_heap, (-a, 'a'))
        if b > 0:
            heapq.heappush(max_heap, (-b, 'b'))
        if c > 0:
            heapq.heappush(max_heap, (-c, 'c'))
        
        while max_heap:
            cnt_curr, ch_curr = heapq.heappop(max_heap)  # Pop the character with max frequency
            cnt_curr = -cnt_curr  # Convert back to positive for use
            
            if len(res) < 2 or res[-1] != ch_curr or res[-2] != ch_curr:
                res.append(ch_curr)  # Append the character to the result
                cnt_curr -= 1  # Decrease frequency
                if cnt_curr > 0:
                    heapq.heappush(max_heap, (-cnt_curr, ch_curr))  # Reinsert if still has occurrences
            else:
                if not max_heap:
                    break  # No other characters available

                cnt_next, ch_next = heapq.heappop(max_heap)  # Next character with max frequency
                cnt_next = -cnt_next  # Convert to positive
                
                res.append(ch_next)  # Append the next character
                cnt_next -= 1  # Decrease frequency
                if cnt_next > 0:
                    heapq.heappush(max_heap, (-cnt_next, ch_next))  # Reinsert next character
                
                heapq.heappush(max_heap, (-cnt_curr, ch_curr))  # Reinsert the original character

        return ''.join(res)
